scripts: add complexity calculation script

MasterToney opened the pull request:
# Purpose

This will add a simple ruby script which calculates the complexity of a lib elektra namespace based on the number and type of found configuration keys.
The ruby swig binding is required for this script to work.

# Checklist

Check relevant points but please do not remove entries.
For docu fixes, spell checking, and similar nothing
needs to be checked.

- [x] commit messages are fine ("module: short statement" syntax and refer to issues)
- [ ] I added unit tests
- [ ] I ran all tests locally and everything went fine
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see doc/CODING.md)
- [ ] meta data is updated (e.g. README.md of plugins)
- [ ] release notes are updated (doc/news/_preparation_next_release.md)

@markus2330 please review my pull request

markus2330 commented on scripts/complexity.rb:
```diff
@@ -0,0 +1,100 @@
+# @author Anton Hößl <e1427045@student.tuwien.ac.at>
```
Please add a `#! interpreter line`, make it executable, and install the file.

markus2330 commented on scripts/complexity.rb:
```diff
@@ -0,0 +1,100 @@
+# @author Anton Hößl <e1427045@student.tuwien.ac.at>
+# @brief This script approximates the complexity of a elektra configuration namespace
```
"a elektra configuration namespace" -> "Elektra configuration settings"

markus2330 commented on scripts/complexity.rb:
```diff
@@ -0,0 +1,100 @@
+# @author Anton Hößl <e1427045@student.tuwien.ac.at>
+# @brief This script approximates the complexity of a elektra configuration namespace
+# @date 18.04.2018
+# @tags generator
```
why generator?

markus2330 replied:
You are not bound to the tags which currently exist. Feel free to use any tag you find useful.

E.g. "specification" and "calculation" might be useful tags.

markus2330 commented on scripts/complexity.rb:
```diff
@@ -0,0 +1,100 @@
+# @brief This script approximates the complexity of a elektra configuration namespace
+# @date 18.04.2018
+# @tags generator
+
```
Please add short description of how to use the script.

markus2330 commented on scripts/complexity.rb:
```diff
@@ -0,0 +1,100 @@
+require 'kdb'
+
+# function checks if the type of the key is known and returns the number of bits needed to represent all possible values
+# of that key, returns 1 if it is a unknown type. Doesn't consider range checks etc...
```
Better to say what it supports (maybe in the short description above)

markus2330 commented on scripts/complexity.rb:
```diff
@@ -0,0 +1,100 @@
+  # get all config keys under our application settings path
+  ret = db.get ks, APP_NS
+rescue Kdb::KDBException
+  puts "could not get Keys under %s: %s" % [APP_NS, $!]
```
under -> below

markus2330 commented:
Thank you for creating this PR. It looks great.

@sanssecours do want to review the Ruby script?

sanssecours commented on scripts/complexity.rb:
```diff
@@ -0,0 +1,100 @@
+
+APP_NS = ARGV[0]
+
+puts "Welcome to the Ruby Complexity script, calculating complexits for namespace:\n"
```
Typo: `complexits` → `complexity`

sanssecours commented on scripts/complexity.rb:
```diff
@@ -0,0 +1,100 @@
+complexityPower = 0;
+
+# go through all keys and process them
+ks.each { |k|
```
Please use `do`-syntax for multiline blocks.

sanssecours commented on scripts/complexity.rb:
```diff
@@ -0,0 +1,100 @@
+}
+
+puts "\n\n"
+numbers_of.each {|key, value| puts "#{key} occured #{value} times."}
```
Typo: `occured` →  `occurred`

sanssecours commented on scripts/complexity.rb:
```diff
@@ -0,0 +1,100 @@
+
+# function checks if the type of the key is known and returns the number of bits needed to represent all possible values
+# of that key, returns 1 if it is a unknown type. Doesn't consider range checks etc...
+def evaluate_type(t, tc, k)
```
Please use more descriptive variable names. For example, instead of `t` you could use `type` or something similar.

markus2330 commented:
@sanssecours  Thank you for this great review!

MasterToney commented:
Hello and thank you for your feedback.
I addressed all change requests you had , in addition to that i also ran roboCop as suggested and refactored accordingly.
@markus2330 please review my pull request

markus2330 commented:
Thank you for the script!

sanssecours commented on scripts/complexity.rb:
```diff
@@ -0,0 +1,125 @@
+#! interpreter line
```
I think we should add something like `#!/usr/bin/env ruby` instead of `#! interpreter line` here ☺️.
